### DISCUSSION TITLE
How does pest generate parsing codes for different rule types (modifiers)?

### DISCUSSION THREAD

**Discussion by WHMHammer** [pest-parser/pest#864] [Q&A]
I am trying to understand the following piece of codes for https://github.com/pest-parser/pest/issues/520: https://github.com/pest-parser/pest/blob/a4c93a7e8778007a2610243a19d688de2e6ee4bd/generator/src/generator.rs#L279-L329

The bodies of the generated methods are chains of function calls as listed below:
- Normal: `rule` -> `#expr`
- Silent: `#expr`
- Atomic: `rule` -> `atomic` -> `#expr`
- CompoundAtomic -> `atomic` -> `rule` -> `#expr`
- NonAtomic -> `atomic` -> `rule` -> `#expr`

What makes me wonder is that for `Atomic`, `atomic` is called inside `rule`, but for `CompoundAtomic` and `NonAtomic`, `rule` is called inside `atomic`. Is there a reason that the calling sequence for `Atomic` is different from `CompoundAtomic` and `NonAtomic`? Does the calling sequence matter?

Thank you.

**Comment by NoahTheDuke** [comment]
I don't know the history. Have you used `git blame` to find out when and how those were added? That might shed some light.

**Reply by WHMHammer** [reply]
Thank you @NoahTheDuke, the change history is really interesting.

In [the very first version](https://github.com/pest-parser/pest/commit/4728f1bccab23752449ab44d47d637ad9903da62?diff=split#diff-8bf623bd40912c5af526946a30639065da0f07cde1022df11add057d9dd2fdc8R147-R172) of the `generate_rule` function, although the Compound Atomic modifier wasn't there, the calling sequences of the Atomic and Non-Atomic rules are the same: they are both `atomic` -> `rule` -> `#expr`.

In [a later commit](https://github.com/pest-parser/pest/commit/a76b989f7f625cb5fa660fd9b907864015aca040#diff-8bf623bd40912c5af526946a30639065da0f07cde1022df11add057d9dd2fdc8R154-R179), the calling sequence in both Atomic and Non-Atomic (Compund Atomic still did not exist yet) were flipped, both becoming `rule` -> `atomic` -> `#expr`. The commit message was `Added tests and fixed some issues.`, but the author didn't specify the details of the issue nor the purpose of changing the call sequence.

And then in [another commit](https://github.com/pest-parser/pest/commit/0ecb0e1a6335e791913812809339934a2ee2815a#diff-8bf623bd40912c5af526946a30639065da0f07cde1022df11add057d9dd2fdc8R165-R190), the Non-Atomic case was removed, and the calling sequence of Compound Atomic changed from `rule` -> `atomic` -> `#expr` back to `atomic` -> `rule` -> `#expr`. The commit message was `Fixed a couple of atomic-related bugs.`, still without any details.

When [Compound Atomic was added back](https://github.com/pest-parser/pest/commit/e7efd971d8b49c4e0d463bbd92f31e5e99b8c59c#diff-8bf623bd40912c5af526946a30639065da0f07cde1022df11add057d9dd2fdc8R165-R203), its calling sequence became `atomic` -> `rule` -> `#expr` as well, different from the sequence of `rule` -> `atomic` -> `#expr` when it was removed in the commit I mentioned in the previous paragraph. The commit message was `Fixed compound atomic's cascading effect.`, without furter details.

I still don't understand the reasons behind these changes yet, but I will continue my research.

Hi @dragostis, you made all 4 commits above. Would you mind kindly explain the decisions and provide me with some guidance on implementing `_@`, `_$`, and `_!` modifiers, as mentioned in [#520](https://github.com/pest-parser/pest/issues/520#issuecomment-1582635365)? I would really appreciate your help! Thank you and thank you for building this wonderful library!

**Reply by tomtau** [reply]
@WHMHammer I think the calling sequence may matter for the state scoping (how the interior rules should behave / whether they produce inner tokens or not).
